core/controllers/android.py

```diff
@@ -343,3 +348,118 @@ def _create_dummy_skill(
             skill_id, skill_description, rubrics)
         skill.update_explanation(state_domain.SubtitledHtml('1', explanation))
         return skill
+
+
+class AndroidActivityHandlerHandlerNormalizedRequestDict(TypedDict):
+    """Dict representation of AndroidActivityHandler's normalized_request
+    dictionary.
+    """
+
+    activity_type: str
+    activity_id: str
+    activity_version: int
+    api_key: str
+
+
+class AndroidActivityHandler(base.BaseHandler[
+    Dict[str, str], AndroidActivityHandlerHandlerNormalizedRequestDict
+]):
+    """Handler for providing activities to Android."""
+
+    GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+
+    URL_PATH_ARGS_SCHEMAS = {
+        'secret': {
+            'schema': {
+                'type': 'basestring'
+            }
+        }
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'GET': {
+            'activity_type': {
+                'schema': {
+                    'type': 'basestring',
+                    'choices': [
+                        constants.ACTIVITY_TYPE_EXPLORATION,
+                        constants.ACTIVITY_TYPE_STORY,
+                        constants.ACTIVITY_TYPE_SKILL,
+                        constants.ACTIVITY_TYPE_SUBTOPIC,
+                        constants.ACTIVITY_TYPE_LEARN_TOPIC,
+                        constants.ACTIVITY_TYPE_CLASSROOM
+                    ]
+                },
+            },
+            'activity_id': {
+                'schema': {
+                    'type': 'basestring'
+                }
+            },
+            'activity_version': {
+                'schema': {
+                    'type': 'int'
+                },
+                'default_value': None
+            }
+        }
+    }
+
+    ACTIVITY_TYPE_TO_FETCHER = {
+        constants.ACTIVITY_TYPE_EXPLORATION: exp_fetchers.get_exploration_by_id,
+        constants.ACTIVITY_TYPE_STORY: story_fetchers.get_story_by_id,
+        constants.ACTIVITY_TYPE_SKILL: skill_fetchers.get_skill_by_id,
+        constants.ACTIVITY_TYPE_SUBTOPIC: (
+            subtopic_page_services.get_subtopic_page_by_id),
+        constants.ACTIVITY_TYPE_LEARN_TOPIC: topic_fetchers.get_topic_by_id
+    }
+
+    # Here, the 'secret' url_path_argument is not used in the function body
+    # because the actual usage of 'secret' lies within the
+    # 'is_from_oppia_android_build' decorator, and here we are getting 'secret'
+    # because the decorator always passes every url_path_args to HTTP methods.
+    @acl_decorators.is_from_oppia_android_build
```

Comment:
Quick question before I delve into this PR further: how does this work locally? When I tried running the web server locally and calling the endpoint with curl it failed to resolve any secrets (I had to disable the secret check in order to test the endpoints locally).

Reply:
Ah, right, I need to write secrets emulator for local usage. I will do that today or tomorrow.

Reply:
@BenHenning Currently, we set the secrets only using Google Cloud Console, we only fetch them using code in Oppia. I think that locally we would have an optional `--secrets` flag for the start script that would accept a dictionary of secrets to be set when starting the local server. These would be stored in an environment variable. Does that make sense? I think this is the fastest way to do this, it might be a bit annoying that you need to restart the server to reset the secrets but I think that is not that bad for this use case.

Reply:
Ah, meant to say this earlier: I really like the idea of using a CLI argument, but it seems (per my other comments & edits to the branch) that this will unfortunately not be an option.